### PR TITLE
ACTIN-953: Only annotate events when generateKnownEvents is set

### PR DESCRIPTION
**Pull request by jbartletthmf**
Replace synchronized list with map and collect.
Refactor using streaming conventions.

### CODE REVIEW

File: algo/src/main/java/com/hartwig/serve/sources/ckb/CkbExtractor.java
Comment (pauldwolfe) [inline algo/src/main/java/com/hartwig/serve/sources/ckb/CkbExtractor.java:135]: Where was the bug? Looks like the flag was checked?

File: algo/src/main/java/com/hartwig/serve/sources/ckb/CkbExtractor.java
Comment (pauldwolfe) [inline algo/src/main/java/com/hartwig/serve/sources/ckb/CkbExtractor.java:110]: If we inlined the `extractAndInterpretEvent` maybe we can save some processing when flag is false and remove the second conditional checking generateKnownEvents?

File: algo/src/main/java/com/hartwig/serve/sources/ckb/CkbExtractor.java
Comment (pauldwolfe) [inline algo/src/main/java/com/hartwig/serve/sources/ckb/CkbExtractor.java:183]: this is the second conditional I was referring to.

File: algo/src/main/java/com/hartwig/serve/sources/ckb/CkbExtractor.java
Comment (jbartletthmf) [reply algo/src/main/java/com/hartwig/serve/sources/ckb/CkbExtractor.java:110]: I extracted that function just because I thought this one was getting a bit long, but happy to refactor. I will say that in general I think we could find more efficient ways to accomplish CKB extraction as there are many places where we replace certain fields with new information that could probably be streamlined. For example, see `CkbVariantAnnotator.annotate` and `ExtractionFunctions.merge`.

File: algo/src/main/java/com/hartwig/serve/sources/ckb/CkbExtractor.java
Comment (jbartletthmf) [reply algo/src/main/java/com/hartwig/serve/sources/ckb/CkbExtractor.java:183]: I'm not sure that inlining is the way to go because this is actually 2 levels deep (`getExtractionResult -> extractAndInterpretEvent -> toExtractionResult`). However, I don't see any reason why we can't inline the annotation at this level and skip the post-extraction annotation step.

File: algo/src/main/java/com/hartwig/serve/sources/ckb/CkbExtractor.java
Comment (pauldwolfe) [reply algo/src/main/java/com/hartwig/serve/sources/ckb/CkbExtractor.java:110]: My point was that `extractAndInterpretEvent` is only called in this one place, and it is the only caller of `toExtractionResult`. So if we only call it when `generateKnownEvents` is true, then you don't need to check that whether `generateKnownEvents` is true on line 183.

Comment (jbartletthmf) [reply algo/src/main/java/com/hartwig/serve/sources/ckb/CkbExtractor.java:110]: `extractAndInterpretEvent` is called for every case where `entry.type() != EventType.UNKNOWN`, so we can't make any assumptions about `generateKnownEvents`.

Comment (pauldwolfe) [reply algo/src/main/java/com/hartwig/serve/sources/ckb/CkbExtractor.java]: I'm a bit confused, looking at the latest changes I think we missed each other. I meant to inline the call to `extractAndInterpretEvent` such that it would only be called if and when the generateKnownEvents was true. I guess I should have said "inline the variable `extraction`" and not the method... 

That said, if you like the way it looks now I think it still solves my concern, the flag is only checked in one place.

**Review by pauldwolfe — APPROVED** [review]

File: algo/src/main/java/com/hartwig/serve/sources/ckb/CkbExtractor.java
Comment (jbartletthmf) [reply algo/src/main/java/com/hartwig/serve/sources/ckb/CkbExtractor.java]: I still don't follow your suggestion...

We needed to call `extractAndInterpretEvent` when `generateKnownEvents` was false too. In that case, the `ExtractionResult` was returned without annotation. That said, annotation now happens with extraction so the extra check is no longer needed.

Comment (pauldwolfe) [reply algo/src/main/java/com/hartwig/serve/sources/ckb/CkbExtractor.java]: I see it now, I'd erroneously read the code as returning Optional.empty when the flag was false. My bad! I think its still nice now though that we don't need to check the flag twice. Alls well that ends well, sorry for any wasted time.

Comment (jbartletthmf) [reply algo/src/main/java/com/hartwig/serve/sources/ckb/CkbExtractor.java]: No problem, I still think this was an improvement.

File: algo/src/test/java/com/hartwig/serve/sources/ckb/CkbVariantAnnotatorTest.java
Comment (kduyvesteyn) [inline algo/src/test/java/com/hartwig/serve/sources/ckb/CkbVariantAnnotatorTest.java:108]: Not a comment but just a remark: Nearly all actual variants have `associatedWithDrugResistance` null, seeing `Y` as default looks weird to me but I guess a test doesn't have to reflect reality :)

Also - FYI: There is no `N` value here, since CKB can never definitively know whether a variant implies drug resistance for any of the existing or future drugs!

File: algo/src/main/java/com/hartwig/serve/sources/ckb/CkbExtractor.java
Comment (kduyvesteyn) [inline algo/src/main/java/com/hartwig/serve/sources/ckb/CkbExtractor.java:122]: Unrelated to your change, but maybe `curateCodons` should take in and return an `ExtractionOutput` such that we don't have to pass these specific codons around anymore.

Also, FYI: We curate codon 600 of BRAF because at that time the 38 ensembl cache used the wrong transcript for BRAF. I assume this is fixed by now (otherwise we'd have problems using OncoPanel 38!), we'd know for sure if we run a BRAF V600E through PAVE on 38 (or alternatively, run COLO pipeline on 38 since COLO has V600E mutation)

**Review by kduyvesteyn — APPROVED** [review]
Looks all good to me, some general comments:
 - `CkbExtractor` doesn't adhere to `serve_code_scheme`. This code scheme should be identical to ACTIN, so you should be able to auto-reformat.
 - You missed notnull annotations at various places. The policy is to annotate every input and output of every function + every property of a class. 
 - I'd suggest doing a regression on this upgrade just to be sure there is no actual change in SERVE output.

**Comment by kduyvesteyn** [comment]
@pauldwolfe it seems like your cloud build is trying to publish a docker image upon each push and fails because it expects a tag? Shouldn't we only publish docker images upon the push of a tag?

**Comment by pauldwolfe** [comment]
> @pauldwolfe it seems like the cloud build is trying to publish a docker image upon each push and fails because it expects a tag? Shouldn't we only publish docker images upon the push of a tag?

I'll have to make a separate PR cloudbuild.yaml that doesn't do the docker stuff. I'll do that this morning.

**Comment by jbartletthmf** [comment]
IntelliJ made the formatting worse when I allowed reformatting before the commit, but calling Reformat directly seemed to fix it.

I'm trying to add all the annotations but I'm spoiled by Kotlin 😛

File: algo/src/main/java/com/hartwig/serve/sources/ckb/CkbExtractor.java
Comment (jbartletthmf) [inline algo/src/main/java/com/hartwig/serve/sources/ckb/CkbExtractor.java:132]: This change makes the output match the current research output, but is it desirable? In master, the input list of codons is initialized for each entry of `actionableEntries` or empty if there are no entries. Omitting input codons when there are no actionable entries reduces the known codon list from 124 to 30 (e.g. 6 codons for gene ERBB2, which is omitted entirely with this change).

File: ckb-importer/src/main/java/com/hartwig/serve/ckb/json/CkbJsonDirectoryReader.java
Comment (jbartletthmf) [inline ckb-importer/src/main/java/com/hartwig/serve/ckb/json/CkbJsonDirectoryReader.java:34]: Parallelizing sped up reading from about 3 minutes to 1 minute on my laptop, which is significant when skipping hotspot resolution.

File: algo/src/main/java/com/hartwig/serve/sources/ckb/CkbExtractor.java
Comment (kduyvesteyn) [reply algo/src/main/java/com/hartwig/serve/sources/ckb/CkbExtractor.java:133]: This means that we only consider a codon to be a known codon if the codon has associated clinical evidence? 

I think we've made that choice since CKB is actionability-focused, and we only want to consider a codon to be known (=pathogenic) if its evidence passes our filtering.

Comment (jbartletthmf) [reply algo/src/main/java/com/hartwig/serve/sources/ckb/CkbExtractor.java:133]: Thanks for confirming!

**Comment by jbartletthmf** [comment]
@pauldwolfe not much has changed here since you gave approval so I'll merge this once I have a successful validation run if you have no objections.

**Comment by pauldwolfe** [comment]
> @pauldwolfe not much has changed here since you gave approval so I'll merge this once I have a successful validation run if you have no objections.

No objections.

**Pull request merged**
